### DISCUSSION TITLE
Custom decoder issue

### DISCUSSION THREAD

**Discussion by pfusik** [ImageMagick/ImageMagick#3046] [Development]
I'm a developer of [RECOIL](http://recoil.sourceforge.net), a decoder of 400+ image formats native to vintage computers.
It includes an ImageMagick decoder which I'm now upgrading to ImageMagick 7: https://sourceforge.net/p/recoil/code/ci/master/tree/imagemagick/recoilmagick.c
I'm testing with `ImageMagick-7.0.10-53-Q16-HDRI-x64-dll.exe`.
With `IM_MOD_RL_recoil_.dll` installed into `C:\Program Files\ImageMagick-7.0.10-Q16-HDRI\modules\coders` it works fine under IMDisplay, but I have a problem with `magick.exe`.
It lists all the formats:

    C:\0\a8\recoil>"\Program Files\ImageMagick-7.0.10-Q16-HDRI\magick.exe" identify -list format | head
       Format  Module    Mode  Description
    -------------------------------------------------------------------------------
          256* RECOIL    r--   Amiga or Atari 8-bit
            3* RECOIL    r--   ZX Spectrum
         3200* RECOIL    r--   Apple IIGS
         3201* RECOIL    r--   Apple IIGS
          32K* RECOIL    r--   Apple IIGS
          3FR  DNG       r--   Hasselblad CFV/H3D39II
          3G2  VIDEO     r--   Media Container
          3GP  VIDEO     r--   Media Container

However it cannot decode them:

    C:\0\a8\recoil>"\Program Files\ImageMagick-7.0.10-Q16-HDRI\magick.exe" KMOSS.3200 kmoss.png
    magick.exe: unable to open module file 'C:\Program Files\ImageMagick-7.0.10-Q16-HDRI\modules\coders\IM_MOD_RL_3200_.dll': No such file or directory @ warning/module.c/GetMagickModulePath/675.
    magick.exe: no decode delegate for this image format `3200' @ error/constitute.c/ReadImage/572.

Is it the case that IMDisplay loads all the coders and `magick.exe` does not? If so, how can I:

1. Specify the coder manually? This doesn't work:

       C:\0\a8\recoil>"\Program Files\ImageMagick-7.0.10-Q16-HDRI\magick.exe" recoil:KMOSS.3200 kmoss.png
       magick.exe: unable to open image 'recoil:KMOSS.3200': No such file or directory @ error/blob.c/OpenBlob/3537.

2. Configure `magick.exe` so that it knows the filename extensions for RECOIL?

**Comment by urban-warrior** [comment]
Stand by.  We're investigating.  It may take a month or two to resolve the issue.

**Reply by pfusik** [reply]
Friendly reminder after three weeks. :)

**Comment by urban-warrior** [comment]
Still don't have a solution, its probably going to take a while.  Stand by...

**Reply by dlemstra** [reply]
You cannot open the `.3200` file directly because ImageMagick cannot guess aliases for the module. This means that you will need to specify the module name. I noticed that you already tried that but it raised an exception for you. I am unable to reproduce that issue with the latest version from main. To test your issue I created a custom coder that prints the file name, opens the blob and returns a 1x1 empty image. This is the output that I am getting:

Without specifying the module name:
```console
I:\issues\im7\3046>magick KMOSS.3200 info:
magick: unable to open module file 'C:\Users\Dirk\AppData\Local\ImageMagick\IM_MOD_DB_3200_.dll': No such file or directory @ warning/module.c/GetMagickModulePath/823.
magick: no decode delegate for this image format `3200' @ error/constitute.c/ReadImage/572.
```

And this is what I get when I specify the module name:
```console
I:\issues\im7\3046>magick test:KMOSS.3200 info:
image_info->filename: KMOSS.3200
test:KMOSS.3200 TEST 1x1 1x1+0+0 16-bit sRGB 5B 0.000u 0:00.001
```

**Comment by pfusik** [comment]
With ImageMagick-7.0.11-Q16-HDRI my coder sees:

    image_info->filename=recoil:KMOSS.3200

Is there something my coder should do to strip the module name or was that recently fixed on main?

**Reply by dlemstra** [reply]
We did not recently changed anything. What is the exact version that you used? I tested this with the code that was contained in the `7.0.11-1` release. I tested this with the coder below. Are you getting the same result with this coder?

```c
/*
  Include declarations.
*/
#include "MagickCore/studio.h"
#include "MagickCore/blob.h"
#include "MagickCore/blob-private.h"
#include "MagickCore/cache.h"
#include "MagickCore/color.h"
#include "MagickCore/color-private.h"
#include "MagickCore/colorspace-private.h"
#include "MagickCore/exception.h"
#include "MagickCore/exception-private.h"
#include "MagickCore/image.h"
#include "MagickCore/image-private.h"
#include "MagickCore/list.h"
#include "MagickCore/magick.h"
#include "MagickCore/memory_.h"
#include "MagickCore/pixel.h"
#include "MagickCore/pixel-accessor.h"
#include "MagickCore/quantum-private.h"
#include "MagickCore/static.h"
#include "MagickCore/string_.h"
#include "MagickCore/module.h"

static Image *ReadTESTImage(const ImageInfo *image_info,ExceptionInfo *exception)
{
  Image
    *image;

  MagickBooleanType
    status;

  printf("image_info->filename: %s\n", image_info->filename);
  image=AcquireImage(image_info,exception);
  status=SetImageExtent(image,1,1,exception);
  return(GetFirstImageInList(image));
}

ModuleExport size_t RegisterTESTImage(void)
{
  MagickInfo
    *entry;

  entry=AcquireMagickInfo("TEST","TEST","Test coder");
  entry->decoder=(DecodeImageHandler *) ReadTESTImage;
  (void) RegisterMagickInfo(entry);
  entry=AcquireMagickInfo("TEST","3200","Test coder");
  entry->decoder=(DecodeImageHandler *) ReadTESTImage;
  (void) RegisterMagickInfo(entry);
  return(MagickImageCoderSignature);
}

ModuleExport void UnregisterTESTImage(void)
{
  (void) UnregisterMagickInfo("3200");
  (void) UnregisterMagickInfo("TEST");
}
```

**Reply by pfusik** [reply]
Thanks for the test coder! My exact installation is `ImageMagick-7.0.11-1-Q16-HDRI-x64-dll.exe`.
Your coder works as expected, until I remove these lines:

    entry=AcquireMagickInfo("TEST","TEST","Test coder");
    entry->decoder=(DecodeImageHandler *) ReadTESTImage;
    (void) RegisterMagickInfo(entry);

With these lines removed, I get the same behavior as in my coder:

    C:\0\a8\recoil>"\Program Files\ImageMagick-7.0.11-Q16-HDRI\magick.exe" test:KMOSS.3200 info:
    image_info->filename: test:KMOSS.3200
    magick.exe: unable to open image 'test:KMOSS.3200': No such file or directory @ error/blob.c/OpenBlob/3537.

Note that "RECOIL" is the name of my project, but not of any of the formats it supports.
Now at least I have a workaround of pretending that "RECOIL" is one of the formats.

I wonder why for the`module:filename` syntax it is `first_image_format:` that gets stripped and not `module:` ?
Where is that code?

**Reply by dlemstra** [reply]
Thanks for your help. We will have a fix for this in place in the next release. I get this with our patch applied but without the `"TEST","TEST"` registration:

```console
I:\issues\im7\3046>magick test:KMOSS.3200 info:
image_info->filename: KMOSS.3200
test:KMOSS.3200 3200 1x1 1x1+0+0 16-bit sRGB 0.000u 0:00.000
```

**Reply by pfusik** [reply]
Great, thank you very much!

**Reply by dlemstra** [reply]
In case you are interested these are the steps that we took:

Initial change:
https://github.com/ImageMagick/ImageMagick/commit/99ce81cfc2f4b62743bfe2e66ebb0bd67bbca8fb

This broke something so we had to revert it:
https://github.com/ImageMagick/ImageMagick/commit/c6c9ffe5c204da44647a68f4ee0a405edab146b6

And here is a new attempt that we are testing now:
https://github.com/ImageMagick/ImageMagick/commit/06b16a00778ea31e56cb9eac76b019d2c30b67a9

**Reply by pfusik** [reply]
Thank you.
